src/devices/src/virtio/vsock/event_handler.rs

```diff
@@ -22,201 +22,198 @@
-    fn process_tx(&mut self) -> bool {
-        debug!("vsock: epoll_handler::process_tx()");
-
-        let mut have_used = false;
-
-        while let Some(head) = self.txvq.pop(&self.mem) {
-            let pkt = match VsockPacket::from_tx_virtq_head(&head) {
-                Ok(pkt) => pkt,
-                Err(e) => {
-                    error!("vsock: error reading TX packet: {:?}", e);
-                    have_used = true;
-                    self.txvq.add_used(&self.mem, head.index, 0);
-                    continue;
-                }
-            };
-
-            if self.backend.send_pkt(&pkt).is_err() {
-                self.txvq.undo_pop();
-                break;
+    fn handle_txq_event(&mut self, event: EpollEvent) -> bool {
+        debug!("vsock: TX queue event");
+
+        let event_set = event.event_set();
+        if event_set != EventSet::IN {
+            warn!("vsock: txq unexpected event {:?}", event_set);
+            return false;
+        }
+
+        let mut raise_irq = false;
+        if let Err(e) = self.queue_events[TXQ_INDEX].read() {
+            error!("Failed to get vsock tx queue event: {:?}", e);
+        } else {
+            raise_irq |= self.process_tx();
+            // The backend may have queued up responses to the packets we sent during
+            // TX queue processing. If that happened, we need to fetch those responses
+            // and place them into RX buffers.
+            if self.backend.has_pending_rx() {
+                raise_irq |= self.process_rx();
             }
+        }
+        raise_irq
+    }
+
+    fn handle_evq_event(&mut self, event: EpollEvent) -> bool {
+        debug!("vsock: event queue event");
 
-            have_used = true;
-            self.txvq.add_used(&self.mem, head.index, 0);
+        let event_set = event.event_set();
+        if event_set != EventSet::IN {
+            warn!("vsock: evq unexpected event {:?}", event_set);
+            return false;
         }
 
-        have_used
+        if let Err(e) = self.queue_events[EVQ_INDEX].read() {
+            error!("Failed to consume vsock evq event: {:?}", e);
+        }
+        false
+    }
+
+    // Helper function to invoke the backend notify() method from different EventHandler
+    // contexts (handler_read, handle_write, etc). Will not be necessary anymore after
+    // switching to the new polly.
+    fn notify_backend(&mut self, event: EpollEvent) -> bool {
+        debug!("vsock: backend event");
+
+        self.backend.notify(event.event_set());
+        // After the backend has been kicked, it might've freed up some resources, so we
+        // can attempt to send it more data to process.
+        // In particular, if `self.backend.send_pkt()` halted the TX queue processing (by
+        // reurning an error) at some point in the past, now is the time to try walking the
+        // TX queue again.
+        let mut raise_irq = false;
+        raise_irq |= self.process_tx();
```

Comment:
`let mut raise_irq = self.process_tx();`

Reply:
Yeah, I initially wrote it that way due to force of habit (and copy/paste) :-s Done